students/kristianjf/lesson09/mailroom_5.py

```diff
@@ -0,0 +1,221 @@
+        self._name = name
+        self._donations = [] if donations is None else donations
+    @property
+    def name(self):
+        return self._name
+    @property
+    def donations(self):
+        return self._donations
+    def donate(self, new_donation):
+        self.donations.append(new_donation)
+    def metrics(self):
+        self.num_d = len(self.donations)
+        self.total_d = sum(self.donations)
+        self.avg_d = self.total_d / self.num_d
+
+class DonorsHandle:
+    DATA_STRUCTURE = [] #[('',[])]
+    DATA_STRUCTURE_DICTS = [] #[{'name': None, 'donation': None}]
+    def __init__(self):
+        pass
+    def add_donor(self, donor_object):
+        self.DATA_STRUCTURE.append((donor_object.name, donor_object.donations))
+        self.DATA_STRUCTURE_DICTS.append({'name': donor_object.name, \
+                                    'donation': donor_object.donations})
+
+    def thank_you(self, full_name='', donation_amount=int()):
+        '''
+        If the user (you) selects ‘Send a Thank You’, prompt for a Full Name.
+            If the user types a name not in the list, add that name to the data structure and use it.
+            If the user types a name in the list, use it.
+        Once a name has been selected, prompt for a donation amount.
+            Turn the amount into a number – it is OK at this point for the program to crash if \
+            someone types a bogus amount.
+        Once an amount has been given, add that amount to the donation history of the selected user.
+        Finally, use string formatting to compose an email thanking the donor for their generous \
+        donation. Print the email to the terminal and return to the original prompt.
+        '''
+        '''
+        Collect, Format and Validate Full name
+        '''
+        #full_name = ''
+        while not full_name:
+            full_name_input = input('Please enter full name of the recipient: ').split()
+            full_name = self.name_response_valid(full_name_input)
+        '''
+        Process Full Name
+        '''
+        ds_names = [name for name, donation in self.DATA_STRUCTURE]
+
+        if full_name in ds_names:
+            full_name_index = ds_names.index(full_name)
+            print(f'Found {full_name} in data_structure')
+            while not donation_amount:
+                donation_amount = int(input('Please enter a donation amount: '))
+            self.DATA_STRUCTURE[full_name_index][1].append(donation_amount)
+            print(self.output_string().format(name=full_name, donation=donation_amount))
+        else:
+            while not donation_amount:
+                donation_amount = int(input('Please enter a donation amount: '))
+            self.DATA_STRUCTURE.append((full_name, [donation_amount]))
+            print(self.output_string().format(name=full_name, donation=donation_amount))
+        return self.output_string().format(name=full_name, donation=donation_amount)
+
+    def list_donors(self):
```

Comment:
excelent -- stand alone functionality in the right class :-)

don't forget to keep your docstrings up to date ...